### PR TITLE
Fix #7248: Address ConnectionString parsing in AppConfiguration

### LINKED ISSUE #7248: Address ConnectionString parsing in AppConfiguration
Per TODOs in ConfigurationClient_private.cs:

- // TODO (pri 2): this allows elements in the connection string to be in varied order. Should we disallow it?
- // TODO (pri 2): this parser will succeed even if one of the elements is missing (e.g. if cs == "a;b;c". We should fix that.
- // TODO (pri 2): this might throw an obscure exception. Should we throw a consistent exception when the parser fails?

Finalize and implement.  
- Is varied order allowed by connection string format?  
- Fix the second issue.
- Wrap exception and include original exception as inner exception.  Vet with API folks.

Initial research on connection string:
https://azure.microsoft.com/en-in/blog/windows-azure-web-sites-how-application-strings-and-connection-strings-work/
https://blogs.msdn.microsoft.com/kaushal/2017/05/04/azure-app-service-how-to-read-parse-connection-strings-via-php/

### PR DESCRIPTION
**Pull request by AlexanderSher**

### CODE REVIEW

File: sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient_private.cs [hunk 1]
```diff
@@ -50,41 +50,12 @@ private static void ParseConnectionString(string connectionString, out Uri uri,
...
-                else if (segment.StartsWith(secretString, StringComparison.OrdinalIgnoreCase))
-                {
-                    var secretBase64 = segment.Substring(segment.IndexOf('=') + 1);
-                    // TODO (pri 2): this might throw an obscure exception. Should we throw a consisten exception when the parser fails?
-                    secret = Convert.FromBase64String(secretBase64);
-                }
-            };
+            uri = new Uri(parsed.GetRequired("Endpoint"));
+            credential = parsed.GetRequired("Id");
+            // TODO (pri 2): this might throw an obscure exception. Should we throw a consistent exception when the parser fails?
```
Comment (annelo-msft) [inline sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient_private.cs]: Let's take this comment out.  The fix should address it, or not, but changing the exception would be a breaking change that we'd like to avoid unless there's a compelling business justification for it.

File: sdk/core/Azure.Core/src/Azure.Core.csproj [hunk 2]
```diff
@@ -28,13 +28,13 @@
...
   <ItemGroup>
-    <Compile Remove="Shared\*.cs"/>
-    <Compile Include="Shared\Argument.cs"/>
-    <Compile Include="Shared\EventSourceEventFormatting.cs"/>
-    <Compile Include="Shared\HashCodeBuilder.cs"/>
-    <Compile Include="Shared\NullableAttributes.cs"/>
-    <Compile Include="Shared\ContentTypeUtilities.cs"/>
-    <Compile Include="Shared\ClientDiagnostics.cs"/>
-    <Compile Include="Shared\DiagnosticScope.cs"/>
+    <Compile Remove="Shared\*.cs" />
```
Comment (annelo-msft) [inline sdk/core/Azure.Core/src/Azure.Core.csproj]: Do we need this change to Azure.Core?

File: sdk/core/Azure.Core/tests/RequestUriBuilderTest.cs [hunk 3]
```diff
@@ -6,313 +6,313 @@
...
-            var uriBuilder = new RequestUriBuilder();
-            uriBuilder.Reset(new Uri("http://localhost/"));
-            uriBuilder.AppendQuery("a", "b");
-
-            Assert.AreEqual("http://localhost/?a=*", uriBuilder.ToString(new string[]
-            {
-            }, "*"));
-        }
-    }
+//    public class RequestUriBuilderTest
```
Comment (annelo-msft) [inline sdk/core/Azure.Core/tests/RequestUriBuilderTest.cs]: Was it intentional to leave this commented out?

File: sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient_private.cs (hunk 1 above)
Comment (AlexanderSher) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient_private.cs]: Fix doesn't address it cause the problem happens when base64 decode fails. I can wrap it with exception of the same type, but different message.

File: sdk/core/Azure.Core/src/Azure.Core.csproj (hunk 2 above)
Comment (AlexanderSher) [reply sdk/core/Azure.Core/src/Azure.Core.csproj]: No. And I have no idea why did it happen :)

File: sdk/core/Azure.Core/tests/RequestUriBuilderTest.cs (hunk 3 above)
Comment (AlexanderSher) [reply sdk/core/Azure.Core/tests/RequestUriBuilderTest.cs]: No, I thought I've reverted this change. These tests have been messing up with test runner.

File: sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationRecordedTestSanitizer.cs [hunk 4]
```diff
@@ -1,21 +1,21 @@
...
 
+using Azure.Core;
 using Azure.Core.Testing;
 
 namespace Azure.Data.AppConfiguration.Tests
 {
     public class ConfigurationRecordedTestSanitizer : RecordedTestSanitizer
     {
-        public override void SanitizeConnectionString(ConnectionString connectionString)
+        public override string SanitizeConnectionString(string connectionString)
```
Comment (pakrym) [inline sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationRecordedTestSanitizer.cs]: What's the reason for this change?

Idea of this methods signature was that you can call base with the same instance of `ConnectionString` and combine the effect. Now every implementor needs to parse/ToString.

File: sdk/core/Azure.Core/tests/Azure.Core.Tests.csproj [hunk 5]
```diff
@@ -22,6 +22,7 @@
   </ItemGroup>
   <ItemGroup>
     <Compile Include="..\src\Shared\ArrayBufferWriter.cs" />
+    <Compile Include="..\src\Shared\ConnectionString.cs" Link="TestFramework\ConnectionString.cs" />
```
Comment (pakrym) [inline sdk/core/Azure.Core/tests/Azure.Core.Tests.csproj]: Remove the link attribute please.

File: sdk/core/Azure.Core/src/Shared/ConnectionString.cs [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+                while (end > start && char.IsWhiteSpace(str[end - 1]))
+                {
+                    end--;
+                }
+
+                return end - start;
+            }
+        }
+
+        private static bool TryGetNextSegment(in string str, in string separator, ref int start, ref int end)
```
Comment (pakrym) [inline sdk/core/Azure.Core/src/Shared/ConnectionString.cs]: You shouldn't need in for the reference types parameters.

File: sdk/core/Azure.Core/src/Shared/ConnectionString.cs [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+        private static void ValidateWhitespaces(string connectionString, int start, int end, bool allowWhitespaces, string name)
+        {
+            if (allowWhitespaces)
+            {
+                return;
+            }
+
+            if (char.IsWhiteSpace(connectionString[start]))
+            {
+                throw new InvalidOperationException($"Start of the {name} '{connectionString.Substring(start, end - start)}' shouldn't contain spaces.");
```
Comment (pakrym) [inline sdk/core/Azure.Core/src/Shared/ConnectionString.cs]: Do we really want to enforce this? When would trimming the name for customers be wrong?

File: sdk/core/Azure.Core/src/Shared/ConnectionString.cs (hunk 6 above)
Comment (AlexanderSher) [reply sdk/core/Azure.Core/src/Shared/ConnectionString.cs]: yes, but that ensures that I don't change parameters inside method body.

File: sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationRecordedTestSanitizer.cs (hunk 4 above)
Comment (AlexanderSher) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationRecordedTestSanitizer.cs]: `ConnectionString` is an internal type, while `RecordedTestSanitizer.SanitizeConnectionString` is public.

File: sdk/core/Azure.Core/tests/Azure.Core.Tests.csproj (hunk 5 above)
Comment (AlexanderSher) [reply sdk/core/Azure.Core/tests/Azure.Core.Tests.csproj]: Ok

File: sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient_private.cs (hunk 1 above)
Comment (annelo-msft) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient_private.cs]: Let's create an issue to track it then.  Will fixing it be a breaking change?

File: sdk/core/Azure.Core/src/Shared/ConnectionString.cs [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+
+namespace Azure.Core
+{
+    internal sealed class ConnectionString
+    {
+        private readonly Dictionary<string, string> _segments;
+        private readonly string _segmentSeparator;
+        private readonly string _keyValueSeparator;
+
+        public static ConnectionString Parse(string connectionString, string segmentSeparator = ";", string keyValueSeparator = "=", bool allowFlags = false, bool allowEmptyValues = false, bool allowWhitespaces = false)
```
Comment (annelo-msft) [inline sdk/core/Azure.Core/src/Shared/ConnectionString.cs]: What is a flag in this context?

Comment (AlexanderSher) [reply sdk/core/Azure.Core/src/Shared/ConnectionString.cs]: Something like `url=http://my.azure.com;action=login;ignoreSavedData`

File: sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient_private.cs (hunk 1 above)
Comment (AlexanderSher) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient_private.cs]: `Convert.FromBase64String` throws `FormatException`. If you want to wrap it wih `InvalidOperationException`, then I think it could be a breaking change.

File: sdk/core/Azure.Core/src/Shared/ConnectionString.cs (hunk 7 above)
Comment (AlexanderSher) [reply sdk/core/Azure.Core/src/Shared/ConnectionString.cs]: I've seen such requirements couple times in practice, but since it's an internal type, I can remove the redundant check. Up to you.

File: sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient_private.cs (hunk 1 above)
Comment (annelo-msft) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient_private.cs]: Just to make sure we're both talking about the same thing ....  As long as the library is in preview, breaking changes are fine.  Once we GA the library, breaking changes are not allowed.  Keeping this comment in the source code implies that we have a TODO that would introduce a breaking change after GA.  We should either make a decision to do this now, or decide not to and remove the TODO.  It would be worth following the precedent set in other libraries.  Does storage use FromBase64String in their connection string parsing?  If so, do they wrap FormatException in InvalidOperationException?  My recommendation would be to follow their precedent to have consistent behavior across the Azure SDK.

File: sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationRecordedTestSanitizer.cs (hunk 4 above)
Comment (annelo-msft) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationRecordedTestSanitizer.cs]: I'm a little confused here.  @AlexanderSher it looks like you've introduced a new internal class ConnectionString under core/src/Shared, whereas there was already a public ConnectionString type under core/tests/TestFramework.  Would it be possible to unify these two classes?  It's confusing to have two classes with the same name in the same project, and we should have a good reason to do this if we do.

I also want to make sure we address Pavel's comment -- it sounds like changing this method signature violates the intention behind how the method was designed.  Can we find a way to achieve these goals without going against the intended design?  If needed, let's work with Pavel on this.

File: sdk/core/Azure.Core/tests/TestFramework/RecordedTestSanitizer.cs [hunk 9]
```diff
@@ -39,10 +39,7 @@ public virtual byte[] SanitizeBody(string contentType, byte[] body)
             return body;
         }
 
-        public virtual void SanitizeConnectionString(ConnectionString connectionString)
-        {
-        }
-
+        public virtual string SanitizeConnectionString(string connectionString) => connectionString;
```
Comment (annelo-msft) [inline sdk/core/Azure.Core/tests/TestFramework/RecordedTestSanitizer.cs]: Similar to comment above, let's try to avoid changing this method signature unless there's a super compelling reason to do this.  It looks like it's also used in identity and storage, so we'd need to update those libraries as well if we changed this.

File: sdk/core/Azure.Core/src/Shared/ConnectionString.cs [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+        }
+
+        private ConnectionString(Dictionary<string, string> pairs, string segmentSeparator, string keyValueSeparator)
+        {
+            _segments = pairs;
+            _segmentSeparator = segmentSeparator;
+            _keyValueSeparator = keyValueSeparator;
+        }
+
+        public string GetRequired(string key)
```
Comment (annelo-msft) [inline sdk/core/Azure.Core/src/Shared/ConnectionString.cs]: I'm a little worried about introducing terms like "key" and "flag" here (which also show up in the test code), even in an internal type, when these terms are so overloaded.  At a minimum, we should have comments defining them for folks coming to the code for the first time, or folks coming back to the code after they've been away for a while.  But ideally, we'd use more descriptive names.  Segment and pairs I'm fine with.

File: sdk/core/Azure.Core/src/Shared/ConnectionString.cs (hunk 8 above)
Comment (annelo-msft) [inline sdk/core/Azure.Core/src/Shared/ConnectionString.cs]: Is it necessary to make all of these things configurable?  It adds complexity and length to the code (and this increases the bug surface).  If we don't have evidence that these things would show up in a connection string, I'd make this as minimal as needed for now, given it's an internal class and it would be ok to change the API for it later.

Comment (annelo-msft) [reply sdk/core/Azure.Core/src/Shared/ConnectionString.cs]: I'm still not sure I understand what a flag is, and what it would mean to not allow flags.

File: sdk/core/Azure.Core/src/Shared/ConnectionString.cs [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+        public bool HasFlag(string key)
+        {
+            if (!_segments.TryGetValue(key, out var value))
+            {
+                return false;
+            }
+
+            if (value != null)
+            {
+                throw new InvalidOperationException($"Required segment with the key '{key}' isn't a flag and has value '{value}'.");
```
Comment (annelo-msft) [inline sdk/core/Azure.Core/src/Shared/ConnectionString.cs]: This error would be customer-facing, right?  If so, key and flag need to be common terminology around connection strings.  I can't find anything using these words with these meanings in a search on Azure connection strings:

https://docs.microsoft.com/en-us/azure/storage/common/storage-configure-connection-string
https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-get-connection-string
https://docs.microsoft.com/en-us/azure/mysql/howto-connection-string

File: sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient_private.cs (hunk 1 above)
Comment (AlexanderSher) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient_private.cs]: Storage checks if secret key is valid base64 string, and uses `null` if it isn't. We can't do this cause we don't allow `null` values for the secret. But I think the closest would be wrapping the exception with `InvalidOperationException`.

File: sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationRecordedTestSanitizer.cs (hunk 4 above)
Comment (AlexanderSher) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationRecordedTestSanitizer.cs]: > Would it be possible to unify these two classes?

That's exactly what I've done (sorry, I've forgotten to delete `sdk/core/Azure.Core/tests/TestFramework/ConnectionString.cs` in original commit, fixed now).

> it sounds like changing this method signature violates the intention behind the method's design.

To preserve original design, we either need to make `sdk/core/Azure.Core/src/Shared/ConnectionString.cs` public or convert `TestFramework/ConnectionString` into public wrapper and use it in tests. Up to @pakrym .

File: sdk/core/Azure.Core/src/Shared/ConnectionString.cs (hunk 7 above)
Comment (pakrym) [reply sdk/core/Azure.Core/src/Shared/ConnectionString.cs]: I wouldn't put limitations on inputs if they are not essential for the functionality.

File: sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationRecordedTestSanitizer.cs (hunk 4 above)
Comment (pakrym) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationRecordedTestSanitizer.cs]: I see, what you currently have is fine.

File: sdk/core/Azure.Core/src/Shared/ConnectionString.cs (hunk 7 above)
Comment (AlexanderSher) [reply sdk/core/Azure.Core/src/Shared/ConnectionString.cs]: Done.

File: sdk/core/Azure.Core/src/Shared/ConnectionString.cs (hunk 8 above)
Comment (AlexanderSher) [reply sdk/core/Azure.Core/src/Shared/ConnectionString.cs]: Removed

File: sdk/core/Azure.Core/src/Shared/ConnectionString.cs (hunk 11 above)
Comment (AlexanderSher) [reply sdk/core/Azure.Core/src/Shared/ConnectionString.cs]: Fixed.

File: sdk/core/Azure.Core/src/Shared/ConnectionString.cs [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+                if (segment.Value != null)
+                {
+                    stringBuilder.Append(_keyValueSeparator).Append(segment.Value);
+                }
+            }
+
+            return stringBuilder.ToString();
+        }
+
+
```
Comment (annelo-msft) [inline sdk/core/Azure.Core/src/Shared/ConnectionString.cs]: nit: looks like there's an extra carriage return here.  Are static analyzers running on this?

File: sdk/core/Azure.Core/src/Shared/ConnectionString.cs [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+            while (TryGetNextSegment(connectionString, segmentSeparator, ref segmentStart, ref segmentEnd))
+            {
+                var kvSeparatorIndex = connectionString.IndexOf(keyValueSeparator, segmentStart, segmentEnd - segmentStart, StringComparison.Ordinal);
+                int keyStart = GetStart(connectionString, segmentStart);
+                int keyLength = GetLength(connectionString, keyStart, kvSeparatorIndex);
+
+                var key = connectionString.Substring(keyStart, keyLength);
+                if (segments.ContainsKey(key))
+                {
+                    throw new InvalidOperationException($"Duplicated key '{key}'");
```
Comment (annelo-msft) [inline sdk/core/Azure.Core/src/Shared/ConnectionString.cs]: I'm still concerned about using "key" in a public-facing exception.  In most of the connection string documentation, key is referring to an auth key.

Comment (AlexanderSher) [reply sdk/core/Azure.Core/src/Shared/ConnectionString.cs]: Renamed to `keyword`.

File: sdk/core/Azure.Core/src/Shared/ConnectionString.cs (hunk 10 above)
Comment (AlexanderSher) [reply sdk/core/Azure.Core/src/Shared/ConnectionString.cs]: Renamed to `keyword`

File: sdk/core/Azure.Core/src/Shared/ConnectionString.cs (hunk 12 above)
Comment (AlexanderSher) [reply sdk/core/Azure.Core/src/Shared/ConnectionString.cs]: Fixed

**Review by annelo-msft — APPROVED** [review]

File: sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient_private.cs (hunk 1 above)
Comment (AlexanderSher) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient_private.cs]: Wrapped with `InvalidOperationException`
